contracts/OlasMech.sol

```diff
@@ -90,118 +87,91 @@ abstract contract OlasMech is Mech, IErrorsMech, ImmutableStorage {
     }
 
     /// @dev Performs actions before the delivery of a request.
-    /// @param requester Requester address.
     /// @param requestId Request Id.
     /// @param data Self-descriptive opaque data-blob.
     /// @return requestData Data for the request processing.
-    function _preDeliver(
-        address requester,
-        uint256 requestId,
-        bytes memory data
-    ) internal virtual returns (bytes memory requestData);
+    function _preDeliver(uint256 requestId, bytes memory data) internal virtual returns (bytes memory requestData);
 
     /// @dev Registers a request.
     /// @param requester Requester address.
-    /// @param requestId Request Id.
-    /// @param data Self-descriptive opaque data-blob.
+    /// @param requestIds Set of request Ids.
+    /// @param datas Set of corresponding self-descriptive opaque data-blobs.
     function _request(
         address requester,
-        uint256 requestId,
-        bytes memory data
+        uint256[] memory requestIds,
+        bytes[] memory datas
     ) internal virtual {
-        mapUndeliveredRequestsCounts[requester]++;
-        // Record the requestId => sender correspondence
-        mapRequestAddresses[requestId] = requester;
-
-        // Record the request Id in the map
-        // Get previous and next request Ids of the first element
-        uint256[2] storage requestIds = mapRequestIds[0];
-        // Create the new element
-        uint256[2] storage newRequestIds = mapRequestIds[requestId];
-
-        // Previous element will be zero, next element will be the current next element
-        uint256 curNextRequestId = requestIds[1];
-        newRequestIds[1] = curNextRequestId;
-        // Next element of the zero element will be the newly created element
-        requestIds[1] = requestId;
-        // Previous element of the current next element will be the newly created element
-        mapRequestIds[curNextRequestId][0] = requestId;
+        uint256 numRequests = requestIds.length;
+
+        for (uint256 i = 0; i < requestIds.length; ++i) {
+            uint256 requestId = requestIds[i];
+
+            // TODO optimize or remove?
```

Comment:
Still thinking it's a good idea to register all requests on the side of a mech, as this way mechs will not be able to deliver requests they have not registered by the marketplace (except for requests with signatures).

Reply:
can they not look up the same on the mm?

Reply:
They could, one could go through events though and filter by mechs, where in the current version each mech holds their full record of undelivered requests.